### PR TITLE
Reworked function types. Added haxe 4 function syntax support

### PR DESCRIPTION
**Pull request by mayakwd**

### CODE REVIEW

**Comment by EricBishton** [comment]
Hi Ilya,  This will take me two days' time to review.  I won't be able to start until at least Wednesday.  (Hot-fix also must come first.)  I'm just letting you know what to expect.

**Comment by mayakwd** [comment]
Found some issues, need more work to be done

**Comment by mayakwd** [comment]
Mixed function types support were dropped after explanations by @nadako.

**Comment by EricBishton** [comment]
@mayakwd Are you finished and ready for a review?

**Comment by mayakwd** [comment]
@EricBishton, yes, it's finished.

File: testData/annotation.semantic/ArrayAssignmentBadArrowFunction.hx [hunk 1]
```diff
@@ -1,7 +1,7 @@
 package;
 class Test {
-  <error descr="Incompatible type Array<Int -> Int> can't be assigned from Array<Int -> Float = 1.0>">var should_warn3: Array<Int->Int> = [ a -> 1.0 ];</error>
+  <error descr="Incompatible type Array<(Int) -> Int> can't be assigned from Array<(Int) -> Float>">var should_warn3: Array<Int->Int> = [ a -> 1.0 ];</error>
```
[the remark this answers is not in the document]
Comment (nadako) [reply testData/annotation.semantic/ArrayAssignmentBadArrowFunction.hx]: Even though we might eventually deprecate the old syntax, `Int->Float` for single-argument functions stays, as it's consistent with arrow function syntax too.

**Pull request merged**